### PR TITLE
HUB-185: Add Resume Links to Pause and Resume

### PR DESCRIPTION
**Pull request by chrisclayson**
Add new endpoint to PausedRegistrationController to start a new journey based on IDP name passed in URL.

### CODE REVIEW

File: app/controllers/partials/journey_hinting_partial_controller.rb [hunk 1]
```diff
@@ -16,6 +16,21 @@ def success_entity_id
     journey_hint_value.nil? ? nil : journey_hint_value['SUCCESS']
   end
 
+  def magic_link?
+    journey_hint = journey_hint_value
+    ! (journey_hint.nil? || journey_hint.fetch('MAGIC', nil).nil?)
```
Comment (jakubmiarka) [inline app/controllers/partials/journey_hinting_partial_controller.rb]: I wouldn't use `MAGIC` but something like 'assisted_resume' maybe? I'll have a look at the rest of the PR next week ;)

File: spec/features/user_visits_magic_link_paused_page_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  context 'and the IDP specified in the URL is valid' do
+    it 'renders page with a list of available services' do
+      stub_transaction_details
+      set_session_and_session_cookies!
+      visit '/paused/stub-idp-one'
+
+      expect(page).to have_content(t('hub.paused_registration.from_magic.heading', idp_name: idp_display_name))
+    end
+  end
```
Comment (jakubmiarka) [inline spec/features/user_visits_magic_link_paused_page_spec.rb]: I would add another context if the IDP specified in the URL is invalid

File: app/controllers/paused_registration_controller.rb [hunk 3]
```diff
@@ -59,9 +76,14 @@ def session_is_valid?
...
-    unless last_idp_value.nil?
-      retrieve_decorated_singleton_idp_array_by_entity_id(current_identity_providers_for_sign_in, last_idp_value).first
+    magic_link_idp_value = magic_link_idp
+    if magic_link_idp_value.nil?
+      last_idp_value = last_idp
+      unless last_idp_value.nil?
+        retrieve_decorated_singleton_idp_array_by_entity_id(current_identity_providers_for_sign_in, last_idp_value).first
+      end
+    else
+      retrieve_decorated_singleton_idp_array_by_simple_id(current_identity_providers_for_sign_in, magic_link_idp_value).first
```
Comment (jakubmiarka) [inline app/controllers/paused_registration_controller.rb]: just wondering if we could simplify this, getting a bit complex and less readable when we mix `if`, `unless`, `else`... maybe move the last_idp_value check to the same line with magic_link_idp_value?

File: spec/features/user_visits_magic_link_paused_page_spec.rb (hunk 2 above)
Comment (chrisclayson) [reply spec/features/user_visits_magic_link_paused_page_spec.rb]: That scenario already covered in unit test ... it tests for the rendering of the different template, which contains no interpolation.

File: config/locales/en.yml [hunk 4]
```diff
@@ -541,6 +542,17 @@ en:
             <li>bookmark this page</li>
           </ul>
           <p>You'll need the email or username and password you used to create an account.</p>
+      from_magic:
+        title: "Verification Paused %{idp_name}"
```
[the remark this answers is not in the document]
Comment (chrisclayson) [reply config/locales/en.yml]: Have posed the question to Sarah.

File: app/controllers/partials/journey_hinting_partial_controller.rb [hunk 5]
```diff
@@ -16,6 +16,16 @@ def success_entity_id
     journey_hint_value.nil? ? nil : journey_hint_value['SUCCESS']
   end
 
+  def resume_link?
+    journey_hint = journey_hint_value
+    ! (journey_hint.nil? || journey_hint.fetch('RESUMELINK', nil).nil?)
```
Comment (jakubmiarka) [inline app/controllers/partials/journey_hinting_partial_controller.rb]: I think codacy complained about the space between `!` and `(`

File: spec/controllers/paused_registration_controller_spec.rb [hunk 6]
```diff
@@ -66,6 +67,49 @@
     end
   end
 
+  context 'user is directed to a magic link from an IDP e-mail with a valid IDP simple_id' do
```
Comment (jakubmiarka) [inline spec/controllers/paused_registration_controller_spec.rb]: I would drop the `magic` word from tests too, but 🤷‍♂️

**Pull request merged**